src/twisted/internet/test/test_tcp.py

```diff
@@ -2539,3 +2966,197 @@ def test_resolveIPv6(self):
+            self.reservedFD.reserve()
+            self.assertEqual(len(self.reservedFileObjects), 1)
+            self.assertFalse(self.reservedFileObjects[0].closed)
+
+
+    def test_reserveEMFILELogged(self):
+        """
+        If reserving the file descriptor fails because of C{EMFILE},
+        the exception is suppressed but logged and the reservation
+        remains unavailable.
+        """
+        exhauster = _ExhaustsFileDescriptors()
+        self.addCleanup(exhauster.release)
+        exhauster.exhaust()
+
+        self.assertFalse(self.reservedFD.available())
+        self.reservedFD.reserve()
+        self.assertFalse(self.reservedFD.available())
+
+        errors = self.flushLoggedErrors(OSError, IOError)
+        self.assertEqual(len(errors), 1)
+        self.assertEqual(errors[0].value.errno, errno.EMFILE)
+
+
+    def test_reserveRaisesNonEMFILEExceptions(self):
+        """
+        Any exception raised while opening the reserve file that is
+        not an L{OSError} or L{IOError} whose errno is C{EMFILE} is
+        allowed through to the caller.
+        """
+        for errorClass in (OSError, IOError, ValueError):
+            # Note that the ValueError will present the errno as its
+            # value.
+            def failsWith(errorClass=errorClass):
+                raise errorClass(errno.EMFILE + 1, "message")
+
+            reserveFD = _FileDescriptorReservation(failsWith)
+            self.assertRaises(errorClass, reserveFD.reserve)
+
+
+    def test_available(self):
+        """
+        The reservation is available after the file descriptor is
+        reserved.
+        """
+        self.assertFalse(self.reservedFD.available())
+        self.reservedFD.reserve()
+        self.assertTrue(self.reservedFD.available())
+
+
+    def test_enterFailsWithoutFile(self):
+        """
+        A reservation without an open file used as a context manager
+        raises a L{RuntimeError}.
+        """
+        with self.assertRaises(RuntimeError):
+            with self.reservedFD:
+                "Does not run."
+
+    def test_enterClosesFileExitOpensFile(self):
+        """
+        Entering a reservation closes its file for the duration of the
+        context manager's block.
+        """
+        self.reservedFD.reserve()
+        self.assertTrue(self.reservedFD.available())
+        with self.reservedFD:
+            self.assertFalse(self.reservedFD.available())
+        self.assertTrue(self.reservedFD.available())
+
+
+    def test_exitOpensFileOnException(self):
+        """
+        An exception raised within a reservation context manager's
+        block does not prevent the file from being reopened.
+        """
+        class TestException(Exception):
+            """
+            An exception only used by this test.
+            """
+
+        self.reservedFD.reserve()
+        with self.assertRaises(TestException):
+            with self.reservedFD:
+                raise TestException
```

Comment:
Is this test missing an `assertTrue(self.reservedFD.available())` at the end?  It seems like it is otherwise I don't see how it actually verifies the behavior in the docstring.